### DISCUSSION TITLE
Manually enable Rosetta on a Podman machine v4

### DISCUSSION THREAD

**Discussion by tnk4on** [containers/podman#21539] [Show and tell]
Rosetta is now supported in [Podman v5.1.0](https://github.com/containers/podman/releases/tag/v5.1.0).

---
This is a temporary solution until this feature is implemented in Podman v5.
If you would like to try Rosetta on a Podman machine as soon as possible, you can do so in the following way
Git repo: https://github.com/tnk4on/podman-machine-rosetta

> [!WARNING]
> This solution only work Podman machine v4.

## Step

0. `export CONTAINERS_MACHINE_PROVIDER=applehv`
1. `podman machine init`
2. Add settings for rosetta in `podman-machine-default.json`
```
% vi ~/.config/containers/podman/machine/applehv/podman-machine-default.json
```

Add the following code in `"devices": []`
```
    {
     "kind": "rosetta",
     "MountTag": "rosetta",
     "InstallRosetta": true
    }
```
3. `podman machine start`
4. `podman machine os apply quay.io/tnk4on/rosetta --restart`
5. After the Podman machine restart is complete, you can run containers with rosetta
```
% podman machine ssh sudo cat /proc/sys/fs/binfmt_misc/rosetta
enabled
interpreter /mnt/rosetta/rosetta
flags: F
offset 0
magic 7f454c4602010100000000000000000002003e00
mask fffffffffffefe00fffffffffffffffffeffffff

% podman run --rm --arch amd64 ubi9 uname -m
x86_64
```

## How to manually build a custom OCI Image

```
git clone https://github.com/tnk4on/podman-machine-rosetta.git
cd podman-machine-rosetta
podman build -t rosetta .
podman push rosetta quay.io/tnk4on/rosetta
```

**Comment by tnk4on** [comment]
Rosetta support for Podman machine v5 is currently WIP.
There are minor bugs, but soon I will open a PR.

**Comment by tnk4on** [comment]
I opend the PR https://github.com/containers/podman/pull/21670

**Comment by joshuapetryk** [comment]
@tnk4on What version of podman did you get this running on? I am unable to get it working on Podman 4.9.2. `cat /proc/sys/fs/binfmt_misc/rosetta` is saying no such file

**Reply by tnk4on** [reply]
@joshuapetryk Fixed and works on v4.9.4.
Please retry it.

**Reply by joshuapetryk** [reply]
This was working after you posted but today I tried again on a fresh podman machine and rosetta service is failing to start. Did something change? Noticed the package was updated recently on quay.io

Note this line: `Failed to add binary format 'rosetta': Permission denied`

```
root@localhost:~# systemctl status rosetta.service
× rosetta.service - Rosetta mount
     Loaded: loaded (/etc/systemd/system/rosetta.service; enabled; preset: disabled)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Wed 2024-04-10 22:51:20 CDT; 57s ago
    Process: 1475 ExecStartPre=mount -t virtiofs -o context=system_u:object_r:nfs_t:s0 rosetta /mnt (code=exited, status=>
    Process: 1476 ExecStart=/bin/systemctl restart systemd-binfmt.service (code=exited, status=1/FAILURE)
   Main PID: 1476 (code=exited, status=1/FAILURE)
        CPU: 6ms

Apr 10 22:51:20 localhost.localdomain systemd[1]: Starting rosetta.service - Rosetta mount...
Apr 10 22:51:20 localhost.localdomain systemctl[1476]: Job for systemd-binfmt.service failed because the control process >
Apr 10 22:51:20 localhost.localdomain systemctl[1476]: See "systemctl status systemd-binfmt.service" and "journalctl -xeu>
Apr 10 22:51:20 localhost.localdomain systemd[1]: rosetta.service: Main process exited, code=exited, status=1/FAILURE
Apr 10 22:51:20 localhost.localdomain systemd[1]: rosetta.service: Failed with result 'exit-code'.
Apr 10 22:51:20 localhost.localdomain systemd[1]: Failed to start rosetta.service - Rosetta mount.
root@localhost:~# sudo systemctl start rosetta.service
Job for rosetta.service failed because the control process exited with error code.
See "systemctl status rosetta.service" and "journalctl -xeu rosetta.service" for details.

root@localhost:~# systemctl status systemd-binfmt.service [B
× systemd-binfmt.service - Set Up Additional Binary Formats
     Loaded: loaded (/usr/lib/systemd/system/systemd-binfmt.service; static)
    Drop-In: /usr/lib/systemd/system/service.d
             └─10-timeout-abort.conf
     Active: failed (Result: exit-code) since Wed 2024-04-10 22:52:51 CDT; 3s ago
       Docs: man:systemd-binfmt.service(8)
             man:binfmt.d(5)
             https://docs.kernel.org/admin-guide/binfmt-misc.html
             https://www.freedesktop.org/wiki/Software/systemd/APIFileSystems
    Process: 1511 ExecStart=/usr/lib/systemd/systemd-binfmt (code=exited, status=1/FAILURE)
   Main PID: 1511 (code=exited, status=1/FAILURE)
        CPU: 2ms

Apr 10 22:52:51 localhost.localdomain systemd[1]: Starting systemd-binfmt.service - Set Up Additional Binary Formats...
Apr 10 22:52:51 localhost.localdomain systemd-binfmt[1511]: /etc/binfmt.d/rosetta.conf:1: Failed to add binary format 'rosetta': Permission denied
Apr 10 22:52:51 localhost.localdomain systemd[1]: systemd-binfmt.service: Main process exited, code=exited, status=1/FAILURE
Apr 10 22:52:51 localhost.localdomain systemd[1]: systemd-binfmt.service: Failed with result 'exit-code'.
Apr 10 22:52:51 localhost.localdomain systemd[1]: Failed to start systemd-binfmt.service - Set Up Additional Binary Formats.
```

**Reply by tnk4on** [reply]
Thanks for the report!
The cause was a change in the SELinux context of the mount of the rosetta directory. It is now fixed.
https://quay.io/repository/tnk4on/rosetta/manifest/sha256:93fe722fa615d97da63f69152d53f691ae8ea74e4c97e0e6e0c882785e741f82

FYI.
It seems that the SELinux context should be `container_runtime_exec_t` in v4.
Note that in v5 (under development), the SELinux context works with `nfs_t`.
https://github.com/containers/podman-machine-os/pull/7

**Reply by joshuapetryk** [reply]
perfect that worked!

**Comment by renyuneyun** [comment]
I'm encountering this error during `podman machine init`:
```
Downloading VM image: fedora-coreos-39.20240322.2.0-applehv.aarch64.raw.gz: done  
Extracting compressed file: podman-machine-default_fedora-coreos-39.20240322.2.0-applehv.aarch64.raw: done  
Error: could not find "vfkit" in one of {[/opt/podman/qemu/bin $BINDIR/../libexec/podman /usr/local/opt/podman/libexec/podman /opt/homebrew/opt/podman/libexec/podman /opt/homebrew/bin /usr/local/bin /usr/local/libexec/podman /usr/local/lib/podman /usr/libexec/podman /usr/lib/podman] {<nil>}}.  To resolve this error, set the helper_binaries_dir key in the `[engine]` section of containers.conf to the directory containing your helper binaries.
```

Unfortunately I don't quite understand how to set this variable -- I know where to put in `container.conf`, but don't know what value should I set. Any ideas?

Some information:
- New to Podman, with a fresh install
- Executed ``eval `/usr/libexec/path_helper -s` `` without restarting computer / re-log-in (see #18676)

**Reply by joshuapetryk** [reply]
I resolved this "vfkit" error by installing vfkit on the host machine (MacOs in my case)

**Reply by renyuneyun** [reply]
@joshuapetryk Thanks for the tip. I managed to install vfkit from homebrew, and re-tried everything.
<del>
A quick question: did you encounter the following error?
```
$ podman machine start
Starting machine "podman-machine-default"
Error: vfkit exited unexpectedly with exit code 0
```
</del>
<del>
--- 
</del>
<del>
I had a corrupted Podman environment due to OOM while doing this command the first time (after install vfkit). Then I updated Podman Desktop (and Podman) as prompted. Then I manually did the following, which seemed to have restored the environment. But unfortunately the above error appears.

1. Remove (Renamed) `~/.config/containers/podman/machine/`
2. `podman system connection rm podman-machine-default`
3. `podman system connection rm podman-machine-default-root`
</del>

No, I made a mistake in editing `podman-machine-default.json`. I changed an entry instead of adding another element into the array. Doing it the right way made everything work.

**Comment by fr12k** [comment]
Found the location and file where rosetta can be disabled or enabled for podman.

https://github.com/containers/common/blob/main/pkg/config/containers.conf

That will then configure podman to start the vfkit process with or without rosetta device.
https://github.com/crc-org/vfkit/blob/main/doc/usage.md#rosetta

Was not easy to find i would add it to the documentation the question is where and how ?

**Reply by tnk4on** [reply]
There is a part for Rosetta in the MACHINE TABLE of Docs in containers.conf.
https://github.com/containers/common/blob/main/docs/containers.conf.5.md

I think the following is appropriate.
https://github.com/containers/podman/tree/main/docs/tutorials

This presentation of mine may be helpful to you.
https://speakerdeck.com/tnk4on/rosetta-support-podman-v5-dot-1

**Reply by fr12k** [reply]
Thanks for sharing. It will save the time for people searching for how to disable rosetta.
